### ISSUE TITLE
Make the Block type polymorphic

### ISSUE THREAD

**Issue by mdimjasevic** [input-output-hk/cardano-ledger-byron#304]
Based on a discussion with @edsko, @dnadales and @nc6, the `Block` type should be made polymorphic.

To integrate the `cs-blockchain` specification implementation into `ouroboros-consensus` (see https://github.com/input-output-hk/ouroboros-network/issues/168), it would be best if the `Types.Block` type was made polymorphic. In particular, to match its LaTeX counterpart, there should be a Haskell type class that defines projection functions from a block type to fields in it. It should also be parameterised over a cryptographic parameter that determines the signing algorithm.

**Comment by mdimjasevic** [comment]
This issue depends on #305.

**Comment by mdimjasevic** [comment]
Closing in favor of https://github.com/input-output-hk/fm-ledger-rules/issues/238, where it is opened as a new ticket.
